### PR TITLE
wgpu-devel into wgpu-stable (~Milestones 1-4)

### PR DESCRIPTION
**Pull request by guusw**
Created wgpu-stable from devel.

State of branch:
- Mesh rendering
- Features & shaders (C++ only)
- Materials
- Drawable/Material shader parameters (C++ only)
- Drawable/Material texture parameters (C++ only)
- ImGui backend for WebGPU
- Various build fixes and improvements
- Builds for Android/iOS

The ImGui SDL/WebGPU backends are copied from their repo but slightly modified to satisfy wgpu-native/web:
- Use explicit pipeline layout
- Convert shader code to wgsl

Changes to build requirements:
- need to run the `./bootstrap` from a terminal (mingw terminal on windows) (only required once)
- rust nightly toolchain is required (`rustup toolchain install nightly`)
- when adding rust targets, they should be installed for nightly toolchain, e.g: `rustup +nightly target add <name>`

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -1 +1,7 @@
-BasedOnStyle: LLVM
\ No newline at end of file
+BasedOnStyle: LLVM
+IndentWidth: 2
+TabWidth: 2
+AccessModifierOffset: -2
+UseTab: Never
+ColumnLimit: 80
```
Comment (Kryptos-FR) [inline .clang-format:6]: 80 is way too small. Let it be at least 120 or even 160.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -42,8 +42,8 @@ jobs:
           sudo apt-get -y update
           sudo apt-get -y install build-essential git cmake wget ninja-build
           cd emsdk
-          ./emsdk install 3.1.5
-          ./emsdk activate 3.1.5
+          ./emsdk install latest
+          ./emsdk activate latest
```
Comment (Kryptos-FR) [inline .github/workflows/main.yml:46]: I'm tempted to say that we should freeze versions and only update them when needed.

It is already an issue with our rust code: we don't push the `Cargo.lock` file which I think is a mistake: basically it makes CI jobs non-reproducible since on the same commit two attempts might use different minor version of the dependencies. Additionally, that's also why the rust cache is not working well on the CI workflow.

File: src/core/runtime.cpp [hunk 3]
```diff
@@ -1896,7 +1896,7 @@ void error_handler(int err_sig) {
 
 #ifdef _WIN32
 #include "debugapi.h"
-static bool isDebuggerPresent() { return (bool)IsDebuggerPresent(); }
+static bool isDebuggerPresent() { return true; (bool)IsDebuggerPresent(); }
```
Comment (Kryptos-FR) [inline src/core/runtime.cpp:1899]: Quick and dirty test?

File: src/gfx/.clang-format [hunk 4]
```diff
@@ -0,0 +1,7 @@
+BasedOnStyle: LLVM
+IndentWidth: 4
+TabWidth: 4
+AccessModifierOffset: -4
+UseTab: Always
+ColumnLimit: 160
+AlignEscapedNewlines: Left
```
Comment (Kryptos-FR) [inline src/gfx/.clang-format:7]: Why this sudden style change for this folder?

Comment (guusw) [reply src/gfx/.clang-format:7]: Yeah let me fix that and make a PR into devel for that

File: src/extra/gfx/chainblocks_types.hpp [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+};
+
+struct CBMaterial {
+  MaterialPtr material;
+  CBShaderParameters shaderParameters;
+
+  void updateVariables();
+};
+
+constexpr uint32_t VendorId = 'cgfx';
```
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/extra/gfx/chainblocks_types.hpp:48]: we limit to 4 cos of CBTypes enum/objects

File: src/extra/gfx/buffer_vars.hpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+#include <magic_enum.hpp>
+#include <vector>
+
+namespace gfx {
+
+inline MeshVertexAttribute getAttributeVertexFormat(const CBVar &attribute) {
+  switch (attribute.valueType) {
+  case CBType::Float2:
+    return MeshVertexAttribute("", 2, VertexAttributeType::Float32);
+    break;
```
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/extra/gfx/buffer_vars.hpp:13]: remove break, unreachable

File: src/gfx/renderer.cpp [hunk 7]
```diff
@@ -0,0 +1,1010 @@
...
+    WGPUCommandEncoderDescriptor desc = {};
+    WGPUCommandEncoder commandEncoder = wgpuDeviceCreateCommandEncoder(device, &desc);
+
+    WGPURenderPassDescriptor passDesc = {};
+    passDesc.colorAttachmentCount = 1;
+
+    WGPURenderPassColorAttachment mainAttach = {};
+    mainAttach.clearColor = clearColor;
+
+    // FIXME
```
Comment (sinkingsugar) [inline src/gfx/renderer.cpp:605]: WHat to fix? Add more detail :)

File: .clang-format (hunk 1 above)
Comment (guusw) [reply .clang-format]: Solved by #165

File: .github/workflows/main.yml (hunk 2 above)
Comment (guusw) [reply .github/workflows/main.yml]: Froze it to 3.1.6 instead of latest. Created #173 for the Cargo.lock issue

File: src/core/runtime.cpp (hunk 3 above)
Comment (guusw) [reply src/core/runtime.cpp]: My bad, removed 😄

File: src/gfx/.clang-format (hunk 4 above)
Comment (guusw) [reply src/gfx/.clang-format]: Resolved in #165

File: src/gfx/renderer.cpp (hunk 7 above)
Comment (guusw) [reply src/gfx/renderer.cpp:605]: Yeah I'll turn it into a GitHub issue.

File: src/gfx/imgui/imgui_impl_sdl.cpp [hunk 8]
```diff
@@ -0,0 +1,614 @@
+// dear imgui: Platform Backend for SDL2
```
[the remark this answers is not in the document]
Comment (guusw) [reply src/gfx/imgui/imgui_impl_sdl.cpp:1]: This implements all the inputs for ImGui (copied from ImGui with some minor changes)
We used this partially previously but now entirely.

File: src/gfx/imgui/imgui_impl_sdl.h [hunk 9]
```diff
@@ -0,0 +1,34 @@
+// dear imgui: Platform Backend for SDL2
```
[the remark this answers is not in the document]
Comment (guusw) [reply src/gfx/imgui/imgui_impl_sdl.h:1]: See below

File: src/extra/gfx/chainblocks_types.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (guusw) [reply src/extra/gfx/chainblocks_types.hpp:48]: I don't think size is the problem (it's already 32-bits of combinations) Even if you increase the size, I think we might want to switch to some other mechanism eventually where we define namespaces based on a unlimited string (e.g. `VendorId VendorId = "chainblocks.graphics";`)

File: cmake/GenerateC.cmake [hunk 10]
```diff
@@ -0,0 +1,8 @@
+
+function(generate_c_string_array FILE VARNAME CONTENT)
```
Comment (sinkingsugar) [inline cmake/GenerateC.cmake:2]: What's this for?

File: src/gfx/renderer.cpp (hunk 7 above)
Comment (guusw) [reply src/gfx/renderer.cpp:605]: Ah the issue was there, updated the comment to indicate what is missing

File: cmake/GenerateC.cmake (hunk 10 above)
Comment (guusw) [reply cmake/GenerateC.cmake:2]: Left over from when I was including a list of shader include paths for the bgfx shader compiler.
Doesn't seem like we need it anymore so I'll remove it

File: include/utility.hpp [hunk 11]
```diff
@@ -215,12 +216,28 @@ template <class CB_CORE> class TParamVar {
...
+    memset(&other._v, 0, sizeof(CBVar));
+  }
+
   ~TParamVar() {
     cleanup();
     CB_CORE::destroyVar(_v);
   }
 
   void warmup(CBContext *ctx) {
+    cleanup();
```
Comment (sinkingsugar) [inline include/utility.hpp:240]: needed in what scenario?
generally warmup is called on cleaned-up chains strictly.

Comment (sinkingsugar) [reply include/utility.hpp:240]: might be a red flag/bug

File: src/core/CMakeLists.txt [hunk 12]
```diff
@@ -134,10 +134,12 @@ target_compile_options(chainblocks-core-static
   PRIVATE -ffast-math -fno-finite-math-only -funroll-loops
 )
 
-target_compile_options(chainblocks-core-static
-  PUBLIC -Wno-multichar
-  PRIVATE -Wno-missing-field-initializers -Wno-deprecated-copy
-)
+if(NOT MSVC)
+  # target_compile_options(chainblocks-core-static
```
Comment (sinkingsugar) [inline src/core/CMakeLists.txt:138]: what was this? don't we need those?

File: src/extra/gfx.cpp [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+    _features.cleanup();
+  }
+
+  void warmup(CBContext *context) {
+    _step = PipelineStepObjectVar.New();
+    _features.warmup(context);
+  }
+
+  std::vector<FeaturePtr> collectFeatures(const CBVar &input) {
+    std::vector<FeaturePtr> result;
```
Comment (sinkingsugar) [inline src/extra/gfx.cpp:127]: might wanna cache this in the block and clear it every time it's used.

File: src/extra/gfx.cpp [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+    std::vector<Var> featureVars(Var(_features.get()));
```
Comment (sinkingsugar) [inline src/extra/gfx.cpp:128]: same, cache

File: src/extra/gfx.cpp [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+    } else {
+      views.push_back(getDefaultView());
+    }
+    return views;
+  }
+
+  PipelineSteps collectPipelineSteps() {
+    PipelineSteps steps;
+    Var stepsCBVar(_steps.get());
+    std::vector<Var> stepVars(stepsCBVar);
```
Comment (sinkingsugar) [inline src/extra/gfx.cpp:314]: malloc, cache it?

File: src/extra/gfx/chainblocks_types.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/extra/gfx/chainblocks_types.hpp:48]: fyi it's inspired by this https://github.com/theos/sdks/blob/fd931ca723ee46994ad9f73c2f0929d1ab9732ca/iPhoneOS9.3.sdk/System/Library/Frameworks/AudioUnit.framework/Headers/AUComponent.h#L19

File: src/gfx/material.hpp [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+#include <memory>
+
+namespace gfx {
+
+// Container for shader parameters (basic/texture)
+struct MaterialParameters {
+  std::map<std::string, ParamVariant> basic;
+  std::map<std::string, TextureParameter> texture;
+
+  void set(const std::string &key, const ParamVariant &param) { basic.insert_or_assign(key, (param)); }
```
Comment (sinkingsugar) [inline src/gfx/material.hpp:15]: could be string_view maybe? no big deal

File: src/gfx/opt.hpp [hunk 17]
```diff
@@ -0,0 +1,6 @@
+#pragma once
+
+#if defined(__clang__) || defined(__GNUC__)
+#define likely(x) __builtin_expect((x), 1)
```
Comment (sinkingsugar) [inline src/gfx/opt.hpp:4]: hmm we have those in chainblocks.h

File: src/extra/inputs.cpp [hunk 18]
```diff
@@ -41,132 +29,78 @@ struct MousePixelPos : public Base {
...
-        return Var(float(event.motion.xrel) / float(_width), float(event.motion.yrel) / float(_height));
-      } else if (event.type == SDL_WINDOWEVENT && event.window.event == SDL_WINDOWEVENT_RESIZED &&
-                 event.window.windowID == _windowId) {
-        const auto window = reinterpret_cast<SDL_Window *>(_sdlWinVar->payload.objectValue);
-        SDL_GetWindowSize(window, &_width, &_height);
+    int2 windowSize = getWindow().getSize();
+
+    for (auto &event : getMainWindowGlobals().events) {
+      if (event.type == SDL_MOUSEMOTION) {
+        return Var(float(event.motion.xrel) / float(windowSize.x), float(event.motion.yrel) / float(windowSize.y));
```
[the remark this answers is not in the document]
Comment (guusw) [reply src/extra/inputs.cpp:54]: Yes, it doesn't have check since it's handled by the window, getSize always gives the correct value

File: src/tests/gfx_window.edn [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+(def Root (Node))
+(def test-chain
+  (Chain
+   "test-chain"
+   :Looped
+   (GFX.MainWindow
+    :Title "SDL Window" :Width 1280 :Height 720 :Debug false)))
+
+(schedule Root test-chain)
+(run Root timestep)
```
Comment (Kryptos-FR) [inline src/tests/gfx_window.edn:11]: Nit: missing newline at the end of the file.

File: src/tests/gfx_materials.edn [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+     .time (Math.Multiply 1.2) (Math.Add 0.3) (Math.FMod 1.0) (Push "c")
+     .time (Math.Multiply 1.4) (Math.Add 0.6) (Math.FMod 1.0) (Push "c")
+     1.0 (Push "c")
+     .c (ToFloat4) >= .color-var
+     .drawables-0 (GFX.Draw)
+     .drawables-1 (GFX.Draw)
+     (GFX.Render :Steps .render-steps :View .view)))))
+
+(schedule Root test-chain)
+(run Root timestep)
```
Comment (Kryptos-FR) [inline src/tests/gfx_materials.edn:79]: Nit: missing newline at the end of the file.

File: src/gfx/window.cpp [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+int2 Window::getSize() const {
+  int2 r;
+  SDL_GetWindowSize(window, &r.x, &r.y);
+  return r;
+}
+
+Window::~Window() { cleanup(); }
+} // namespace gfx
```
Comment (Kryptos-FR) [inline src/gfx/window.cpp:96]: Nit: missing newline at the end of the file.

File: src/gfx/types.hpp [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+  Rect(int width, int height) : width(width), height(height) {}
+  Rect(int x, int y, int width, int height) : x(x), y(y), width(width), height(height) {}
+
+  int getX1() const { return x + width; }
+  int getY1() const { return x + height; }
+  int2 getSize() const { return int2(width, height); }
+
+  static Rect fromCorners(int x0, int y0, int x1, int y1) { return Rect(x0, y0, x1 - x0, y1 - y0); }
+};
+} // namespace gfx
```
Comment (Kryptos-FR) [inline src/gfx/types.hpp:63]: Nit: missing newline at the end of the file.

File: src/gfx/enums.cpp [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+  case IndexFormat::UInt16:
+    return WGPUIndexFormat_Uint16;
+  case IndexFormat::UInt32:
+    return WGPUIndexFormat_Uint32;
+  default:
+    assert(false);
+    return WGPUIndexFormat_Undefined;
+  }
+}
+} // namespace gfx
```
Comment (Kryptos-FR) [inline src/gfx/enums.cpp:185]: Nit: missing newline at the end of the file.

File: src/core/number_types.cpp [hunk 24]
```diff
@@ -93,7 +93,7 @@ template <NumberType Type> struct TNumberTypeTraits {};
 #define NUMBER_TYPE_TRAITS(_Type, _CType)                                         \
   template <> struct TNumberTypeTraits<_Type> : public NumberTypeTraits {         \
     typedef _CType TInner;                                                        \
-    TNumberTypeTraits<_Type>() {                                                  \
+    TNumberTypeTraits() {                                                         \
```
[the remark this answers is not in the document]
Comment (guusw) [reply src/core/number_types.cpp:96]: Are you talking about the right-aligned \ escapes? this is the format currently used by clang-format so they're added automatically.

File: src/gfx/shader/wgsl_mapping.hpp [hunk 25]
```diff
@@ -0,0 +1,106 @@
+#pragma once
```
Comment (sinkingsugar) [inline src/gfx/shader/wgsl_mapping.hpp:1]: can we have a readme.md doc in this folder explaining how this whole thing works btw.

Also pragma once is not really the best (but should be ok for now), prefer proper define guards as they support all corner cases

File: src/tests/imgui_demo.edn [hunk 26]
```diff
@@ -438,7 +438,7 @@
 (defnode main)
 (defloop ui-chain
   (GFX.MainWindow
-   :Title "UI Chainblocks Example" :Width 1280 :Height 768 :Debug true :Fullscreen false
```
Comment (sinkingsugar) [inline src/tests/imgui_demo.edn:441]: what happened to fullscreen btw?

File: src/tests/gfx_materials.edn [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+
+    ; Create view
+    {:Position (Float3 0 15 10) :Target (Float3 0 0 0)} (Math.LookAt) >= .view-transform
+    (GFX.View :View .view-transform) >= .view)
+   (GFX.MainWindow
+    :Title "SDL Window" :Width 1280 :Height 720 :Debug false
+    :Contents
+    (->
+     .time (Math.Add timestep) >= .time
+     .time (Math.Multiply 0.9) (Math.Add 0.0) (Math.FMod 1.0) (Push "c")
```
Comment (sinkingsugar) [inline src/tests/gfx_materials.edn:69]: can do `>> .c`

File: src/tests/gfx_materials.edn [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+     .time (Math.Multiply 1.2) (Math.Add 0.3) (Math.FMod 1.0) (Push "c")
+     .time (Math.Multiply 1.4) (Math.Add 0.6) (Math.FMod 1.0) (Push "c")
+     1.0 (Push "c")
+     .c (ToFloat4) >= .color-var
```
Comment (sinkingsugar) [inline src/tests/gfx_materials.edn:73]: Shouldn't this be `>` ?

File: src/core/number_types.cpp (hunk 24 above)
[the remark this answers is not in the document]
Comment (Kryptos-FR) [reply src/core/number_types.cpp:96]: Oh yes, I missed that it was inside a macro.

File: include/utility.hpp (hunk 11 above)
Comment (guusw) [reply include/utility.hpp]: I'm using these in places like CBDrawable (The ref-counted instance created by the GFX.Drawable block) to track captured transform/material values.

Turns out I don't need this cleanup though, so I changed it to an assert to make sure it's not accidentally used wrongly.

File: src/tests/imgui_demo.edn (hunk 26 above)
Comment (guusw) [reply src/tests/imgui_demo.edn:441]: Yeah it got removed during the rewrite.
Will include in a later PR but made a note so I don't forget (https://github.com/orgs/fragcolor-xyz/projects/1#card-79476112)

File: src/gfx/shader/wgsl_mapping.hpp (hunk 25 above)
Comment (guusw) [reply src/gfx/shader/wgsl_mapping.hpp:1]: Will create a small document for the shader tree/generator (https://github.com/orgs/fragcolor-xyz/projects/1#card-79476718)

Comment (guusw) [reply src/gfx/shader/wgsl_mapping.hpp:1]: Made a note about header guards so I can work on it later and don't forget. (https://github.com/orgs/fragcolor-xyz/projects/1#card-79476639)

File: src/tests/gfx_materials.edn (hunk 27 above)
Comment (guusw) [reply src/tests/gfx_materials.edn]: Done, thanks

File: src/tests/gfx_materials.edn (hunk 28 above)
Comment (guusw) [reply src/tests/gfx_materials.edn]: Done, thanks

File: src/tests/gfx_window.edn (hunk 19 above)
Comment (guusw) [reply src/tests/gfx_window.edn]: Fixed, and all other occurences

File: src/gfx/opt.hpp (hunk 17 above)
Comment (guusw) [reply src/gfx/opt.hpp]: Copied because I wanted to keep the dependency graph `gfx --> chainblocks-extra` It's no longer used now so I removed it.

File: src/extra/gfx.cpp (hunk 14 above)
Comment (guusw) [reply src/extra/gfx.cpp:128]: I added a `Var::intoVec(std::vector<T>&)` for this one.

File: src/extra/gfx.cpp (hunk 13 above)
Comment (guusw) [reply src/extra/gfx.cpp:127]: Done

File: src/extra/gfx.cpp (hunk 15 above)
Comment (guusw) [reply src/extra/gfx.cpp]: done

**Comment by sinkingsugar** [comment]
Can we make it pass CI before I review? @guusw

**Comment by guusw** [comment]
```
* 3e25944c Make chainblock root build include-able (new changes)
* ebe37866 Replace MaterialParameters::set arg with string_view (old wgpu-devel)
```

Starting from 3e25944c up to the current HEAD I merged some of the build changes made for iOS/Android since they also fix some CI issues
Should be all green now

File: src/tests/test_runtime.cpp [hunk 29]
```diff
@@ -679,7 +679,7 @@ TEST_CASE("CBVar-comparison", "[ops]") {
 
   SECTION("Bytes") {
     std::random_device rd;
-    std::uniform_int_distribution<uint8_t> dist(0, 0x07);
+    std::uniform_int_distribution<uint32_t> dist(0, 0x07);
```
[the remark this answers is not in the document]
Comment (guusw) [reply src/tests/test_runtime.cpp:682]: Apparently it doesn't, or it's not supported everywhere.
https://stackoverflow.com/questions/31460733/why-arent-stduniform-int-distributionuint8-t-and-stduniform-int-distri

File: .github/workflows/main.yml [hunk 30]
```diff
@@ -40,32 +40,32 @@ jobs:
       - name: Set up dependencies
         run: |
           sudo apt-get -y update
-          sudo apt-get -y install build-essential git cmake wget ninja-build
+          sudo apt-get -y install build-essential autoconf git cmake wget ninja-build
```
[the remark this answers is not in the document]
Comment (satendrapalsingh) [reply .github/workflows/main.yml:43]: Noted - todo as part of #228

File: .github/workflows/main.yml [hunk 31]
```diff
@@ -755,9 +825,9 @@ jobs:
       - name: Set up dependencies
         run: |
           sudo xcode-select --switch /Applications/Xcode.app
-          brew install boost cmake ninja clang-format
-          brew install openssl@1.1
-          rustup update
+          brew install cmake ninja clang-format automake libtool
+          ./bootstrap
+          rustup toolchain install nightly
```
[the remark this answers is not in the document]
Comment (satendrapalsingh) [reply .github/workflows/main.yml:830]: Noted - todo as part of #228

File: .github/workflows/main.yml [hunk 32]
```diff
@@ -415,8 +447,19 @@ jobs:
         env:
           RUSTUP_USE_CURL: 1
         run: |
-          rustup update
-          rustup default stable-${{ matrix.arch }}-pc-windows-gnu
+          rustup toolchain install nightly
+          rustup +nightly target add x86_64-pc-windows-gnu
```
[the remark this answers is not in the document]
Comment (guusw) [reply .github/workflows/main.yml:451]: Oops, fixed

File: src/gfx/mesh.hpp [hunk 33]
```diff
@@ -0,0 +1,88 @@
+
+#pragma once
```
Comment (sinkingsugar) [inline src/gfx/mesh.hpp:2]: I think we should convert to traditional header guards and stick with them for consistency

File: src/gfx/moving_average.hpp [hunk 34]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+float MovingAverage::getAverage() const {
+  if (length == 0)
+    return 0.0f;
+  float sum = 0.0f;
+  for (int i = 0; i < length; i++) {
+    sum += buffer[i];
+  }
+  return sum / float(length);
```
Comment (sinkingsugar) [inline src/gfx/moving_average.hpp:47]: NaN

File: src/mal/CMakeLists.txt [hunk 35]
```diff
@@ -51,4 +51,10 @@ if(EMSCRIPTEN)
   else()
     set_target_properties(cbl PROPERTIES SUFFIX "-st.js")
   endif()
+
+  target_link_options(cbl PUBLIC
+    "SHELL:-s MODULARIZE=1"
```
Comment (sinkingsugar) [inline src/mal/CMakeLists.txt:56]: how come we need this now?

**Review by sinkingsugar — APPROVED** [review]
Approved conditionally to some questions/points I raised 👍

File: src/gfx/gfx_wgpu.cpp [hunk 36]
```diff
@@ -0,0 +1,116 @@
+#include "gfx_wgpu.hpp"
+#include "stdio.h"
+
+#ifdef __EMSCRIPTEN__
+#include <emscripten/html5.h>
+#endif
+static void tickAsyncRequests() {
+#ifdef __EMSCRIPTEN__
+  printf("tickAsyncRequests\n");
+  emscripten_sleep(1);
```
[the remark this answers is not in the document]
Comment (guusw) [reply src/gfx/gfx_wgpu.cpp:10]: The only place I need this currently is when requesting the webgpu adapter on browsers, since it has a callback.

I could move it up if you prefer, but in that case the GFX.Window would delay until the adapter is there

File: src/core/blocks/time.cpp [hunk 37]
```diff
@@ -6,6 +6,17 @@
...
 namespace chainblocks {
 namespace Time {
+
+template <typename TDur> inline CBVar DurationToVar(const TDur &dur) {
+  using CountType = decltype(dur.count());
+  static_assert(sizeof(CountType) <= sizeof(int64_t));
+
+  CBVar result{};
+  result.valueType = CBType::Float;
+  result.payload.floatValue = CBFloat(dur.count());
```
[the remark this answers is not in the document]
Comment (guusw) [reply src/core/blocks/time.cpp:16]: I'll fix this, I assumed they were all the same type, but EpochMs seems to be Int, the other ones float

File: src/gfx/moving_average.hpp (hunk 34 above)
Comment (sinkingsugar) [reply src/gfx/moving_average.hpp:47]: ah and it's int, true 👍

File: src/gfx/gfx_wgpu.cpp (hunk 36 above)
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/gfx/gfx_wgpu.cpp:10]: It's more about `emscripten_sleep`, it's a wrapper to some async features that we already use directly via scheduler/coroutines. Can't you just use CB_SUSPEND? It would do the same in a better way.

File: .github/workflows/main.yml [hunk 38]
```diff
@@ -406,6 +434,10 @@ jobs:
             arch: x86_64
             artifact: cbl-win64
     steps:
+      - name: Change git line endings
+        run: |
+          git config --system core.autocrlf false
+          git config --system core.eol lf
```
[the remark this answers is not in the document]
Comment (guusw) [reply .github/workflows/main.yml]: Should not longer be required

File: src/tests/gfx_materials.edn [hunk 39]
```diff
@@ -0,0 +1,79 @@
...
+    4 5 1
+    2 3 6
+    6 3 7]})
+
+(def timestep (/ 1.0 120.0))
+(def Root (Node))
+(def test-chain
+  (Chain
+   "test-chain"
+   :Looped
```
[the remark this answers is not in the document]
Comment (guusw) [reply src/tests/gfx_materials.edn:31]: Applied all shortcuts

File: src/gfx/mesh.hpp (hunk 33 above)
Comment (guusw) [reply src/gfx/mesh.hpp]: Changed them

**Review by sinkingsugar — APPROVED** [review]

**Comment by sinkingsugar** [comment]
Looks good, but those two points I have open would be nice if addressed or put in an issue

File: src/gfx/context.cpp [hunk 40]
```diff
@@ -79,10 +137,10 @@ struct ContextMainOutput {
...
 #if GFX_EMSCRIPTEN
     auto callback = [](double time, void *userData) -> EM_BOOL {
-      spdlog::debug("Animation frame callback {}", time);
+      getLogger()->debug("Animation frame callback {}", time);
       return true;
     };
-    spdlog::debug("Animation frame queued");
+    getLogger()->debug("Animation frame queued");
     emscripten_request_animation_frame(callback, this);
     emscripten_sleep(0);
```
Comment (sinkingsugar) [inline src/gfx/context.cpp:145]: So here, if you have `CBContext` it can definitely be replaced with `CB_SUSPEND` and it will allow the node to process other chains too.
`emscripten_sleep` uses the same async/await facilities internally it's an hack.

Comment (sinkingsugar) [reply src/gfx/context.cpp:145]: If it's not critical code just add a note btw for later review.

Comment (guusw) [reply src/gfx/context.cpp]: Forgot it was used here too, technically not necessary, only in standalone sample code.
Removed it here since the chainblocks main loop should take care of it already if it calls sleep with asyncify

File: src/gfx/gfx_wgpu.cpp (hunk 36 above)
[the remark this answers is not in the document]
Comment (guusw) [reply src/gfx/gfx_wgpu.cpp]: I removed this from the context code, requesting device/adapter runs async now so on web it just blank screens until it either gets one or fails during activate.

File: cmake/CBExternalProject.cmake [hunk 41]
```diff
@@ -0,0 +1,196 @@
+include(ExternalProject)
```
[the remark this answers is not in the document]
Comment (guusw) [reply cmake/CBExternalProject.cmake:1]: Check `cb_add_external_project` where I added a comment about usage.
The optional arguments are now documented slightly below that.

**Review by sinkingsugar — APPROVED** [review]
